narwhals/_expression_parsing.py

```diff
@@ -399,3 +357,105 @@ def evaluate_output_names_and_aliases(
             *[(x, alias) for x, alias in zip(output_names, aliases) if x not in exclude]
         )
     return output_names, aliases
+
+
+def operation_is_order_dependent(*args: IntoExpr | Any) -> bool:
+    from narwhals.expr import Expr
+
+    return any(isinstance(x, Expr) and x._metadata["is_order_dependent"] for x in args)
+
+
+class ExprKind(Enum):
+    """Describe which kind of expression we are dealing with.
+
+    Commutative composition rules are:
+    - LITERAL vs LITERAL -> LITERAL
+    - CHANGES_LENGTH vs (LITERAL | AGGREGATION) -> CHANGES_LENGTH
+    - CHANGES_LENGTH vs (CHANGES_LENGTH | TRANSFORM) -> raise
+    - TRANSFORM vs (LITERAL | AGGREGATION) -> TRANSFORM
+    - AGGREGATION vs (LITERAL | AGGREGATION) -> AGGREGATION
+    """
+
+    LITERAL = auto()  # e.g. nw.lit(1)
+    AGGREGATION = auto()  # e.g. nw.col('a').mean()
+    TRANSFORM = auto()  # length-preserving, e.g. nw.col('a').round()
+    CHANGES_LENGTH = auto()  # e.g. nw.col('a').drop_nulls()
+
+
+class ExprMetadata(TypedDict):
+    kind: ExprKind
+    is_order_dependent: bool
+
+
+def combine_metadata(*args: IntoExpr, strings_are_column_names: bool) -> ExprMetadata:
+    # Combine metadata from `args`.
+    from narwhals.expr import Expr
+
+    n_changes_length = 0
+    n_transforms = 0
+    n_aggregations = 0
+    n_literals = 0
+    result_is_order_dependent = False
+
+    for arg in args:
+        if isinstance(arg, str) and strings_are_column_names:
+            n_transforms += 1
+        elif isinstance(arg, Expr):
+            if arg._metadata["is_order_dependent"]:
+                result_is_order_dependent = True
+            kind = arg._metadata["kind"]
+            if kind is ExprKind.AGGREGATION:
+                n_aggregations += 1
+            elif kind is ExprKind.LITERAL:
+                n_literals += 1
+            elif kind is ExprKind.CHANGES_LENGTH:
+                n_changes_length += 1
+            elif kind is ExprKind.TRANSFORM:
+                n_transforms += 1
+            else:  # pragma: no cover
+                msg = "unreachable code"
+                raise AssertionError(msg)
```

Comment:
What do you think about a less branching approach by first partitioning into strings and expressions and then using a counter:

```suggestion
    strings = [arg for arg in args if isinstance(arg, str) and strings_are_column_names]
    exprs = [arg for arg in args if isinstance(arg, Expr)]
    
    result_is_order_dependent = any(expr._metadata["is_order_dependent"] for expr in exprs)
    counter = Counter((expr._metadata["kind"] for expr in exprs))
    
    n_changes_length = counter[ExprKind.CHANGES_LENGTH]
    n_transforms = counter[ExprKind.TRANSFORM] + strings_are_column_names * len(strings)
    n_aggregations = counter[ExprKind.AGGREGATION]
    n_literals = counter[ExprKind.LITERAL]   
```

Reply:
@FBruzzesi one thing stood out from the description

> Theoretically, I'd expect it to make things not-worse, as we're not only looping through args once to construct ExprMetadata, as opposed to 4 times

I do think it is *visually* nicer not branching so much, but you can probably reduce that without reverting to the 4x loops

Reply:
Fair enough, it can probably go down to a single loop + counter, which would be more than what's already here.
Nevermind then

Reply:
Actually, coming back to this and keeping a single loop, we could do:

```py
def combine_metadata(*args: IntoExpr, strings_are_column_names: bool) -> ExprMetadata:
    # combine metadata from `args`
    from narwhals.expr import Expr

    counts = Counter()
    result_is_order_dependent = False

    for arg in args:
        if isinstance(arg, str) and strings_are_column_names:
            counts[ExprKind.TRANSFORM] += 1
        elif isinstance(arg, Expr):
            if arg._metadata["is_order_dependent"]:
                result_is_order_dependent = True
            counts[arg._metadata["kind"]] += 1

    if (
        counts[ExprKind.LITERAL]
        and not counts[ExprKind.AGGREGATION]
        and not counts[ExprKind.TRANSFORM]
        and not counts[ExprKind.CHANGES_LENGTH]
    ):
        return ExprMetadata(kind=ExprKind.LITERAL, is_order_dependent=result_is_order_dependent)

    if counts[ExprKind.CHANGES_LENGTH] > 1:
        msg = "Length-changing expressions can only be used in isolation, or followed by an aggregation"
        raise LengthChangingExprError(msg)
    if counts[ExprKind.CHANGES_LENGTH] and counts[ExprKind.TRANSFORM]:
        msg = "Cannot combine length-changing expressions with length-preserving ones"
        raise ShapeError(msg)

    if counts[ExprKind.CHANGES_LENGTH]:
        result_kind = ExprKind.CHANGES_LENGTH
    elif counts[ExprKind.TRANSFORM]:
        result_kind = ExprKind.TRANSFORM
    else:
        result_kind = ExprKind.AGGREGATION

    return ExprMetadata(kind=result_kind, is_order_dependent=result_is_order_dependent)
```

Reply:
thanks - it's occurred to me that most of these can just be booleans rather than counts, it's just `changes_length` that needs counting, so i've rewritten to just use booleans